shortfin/python/shortfin_apps/sd/components/metrics.py

```diff
@@ -0,0 +1,56 @@
+import logging
+import time
+import asyncio
+from typing import Callable, Any
+
+logger = logging.getLogger(__name__)
+
+# measurement helper
+def measure(fn):
+
+    """
+    Decorator log test start and end time of a function
+    :param fn: Function to decorate
+    :return: Decorated function
+    Example:
+    >>> @timed
+    >>> def test_fn():
+    >>>     time.sleep(1)
+    >>> test_fn()
+    """
+
+    def wrapped_fn(*args: Any, **kwargs: Any) -> Any:
+        start = time.time()
+        ret = fn(*args, **kwargs)
+        duration_str = get_duration_str(start)
+        logger.info(f"Completed {fn.__qualname__} in {duration_str}")
+        return ret
+
+    async def wrapped_fn_async(*args: Any, **kwargs: Any) -> Any:
+        start = time.time()
+        ret = await fn(*args, **kwargs)
+        duration_str = get_duration_str(start)
+        logger.info(f"Completed {fn.__qualname__} in {duration_str}")
+        if fn.__qualname__ == "ClientGenerateBatchProcess.run":
```

Comment:
What's going on with this? Seems like you should just have a kwarg on the decorator to tell it what to do. There's a standard (but tricky) idiom for that using functools.partial...

Reply:
Didn't use partial here but adapted to take a few kwargs. There's still some yuck there where we ping attributes of the decorated method's class to figure out batch size, but still more flexible than before. I may revisit later.